### DISCUSSION TITLE
웹 크롤링이 어떤 방식으로 이뤄지는지 설명해주세요

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#55] [프론트 기술 질문 공간]
https://about.daangn.com/jobs/5622058003/
당근마켓 'Software Engineering Intern, Frontend - 커뮤니티' 포지션 지원 질문을 가져와보았습니다.

**Comment by WooGie911** [comment]
먼저 크롤링(crawling) 이란 Web상에 존재하는 Contents를 수집하는 작업입니다.
쉽게 말해 특정 사이트에서 내가 원하는 정보를 가져오는 행위입니다.

크롤링은 프로그래밍으로 자동화가 가능한데 여러 기법 중 다음과 같은 3가지 기법이 있습니다.
1. HTML 페이지를 가져와서, HTML/CSS등을 파싱하고, 필요한 데이터만 추출하는 기법
2. Open API(Rest API)를 제공하는 서비스에 Open API를 호출해서, 받은 데이터 중 필요한 데이터만 추출하는 기법
3. Selenium등 브라우저를 프로그래밍으로 조작해서, 필요한 데이터만 추출하는 기법

**Comment by michoball** [comment]
파이썬과 같은 프로그램을 이용해서 타겟이 되는 웹사이트에 http 요청을 보내서 html 이나 css 등을 파싱해 필요한 데이터만 가져옵니다. 

아니면 클라이언트 측에서 js를 이용해 동적으로 웹 사이트를 조작해 크롤링을 할 수 있습니다. Selenium, Puppeteer와 같은 도구로 가능합니다. 

앱의 경우에는 Deep Link를 사용해서 특정 컨텐츠나 정보를 추출할 수 있다고 하는데 잘 모릅니다.

**Comment by javascriptbiggosoo** [comment]
웹 페이지를 방문하고, 해당 페이지의 데이터를 추출하며, 그 페이지에 연결된 다른 페이지로 이동하는 과정을 반복하는 자동화된 메커니즘을 말합니다. 
보통 특정 웹사이트에서 데이터를 수집하는데 이용됩니다.
예컨대 크롤링을 활용하면 채용사이트에서 특정 키워드가 포함된 공고를 따로 추출하거나, 이미지를 모아둔 사이트에서 레드벨벳 이미지만 긁어올 수도 있습니다.
추출된 데이터를 취합해 엑셀 파일로 정리하는 등 활용범위가 넓지만, 해당 도메인을 서비스하는 업체에 과도한 트래픽을 발생시키는 문제가 있기 때문에 허용되는 범위 안에서 사용해야합니다.

웹 크롤링은 보통 다음의 단계로 이루어집니다:

- 시작 URL 선택: 크롤링을 시작할 웹페이지의 URL을 선택합니다.
- 웹페이지 요청: 선택한 URL로 HTTP 요청을 보냅니다. 이를 통해 웹 서버는 해당 페이지의 HTML을 반환합니다.
- 웹페이지 파싱: 반환된 HTML을 파싱하여 원하는 데이터를 추출합니다. 이때 HTML을 분석하기 위해 HTML 파서를 사용하며, 파이썬에서는 BeautifulSoup가 대표적인 라이브러리입니다.
- 링크 추출: 파싱 과정에서 새로운 URL을 추출하여 추가적으로 크롤링할 페이지를 찾습니다.
- 반복: 새로운 URL로 위의 과정을 반복합니다.
